Canceling writedelay on double CTRL-C

geekodour opened the pull request:
Attempt for: https://github.com/neovim/neovim/issues/8200

Does not currently solve the issue, wip.

This is an attempt to handle `CTRL-C` gracefully when a quickfix command is to be cancelled. The command is cancelled but the delays set by the quickfix remains.

geekodour committed:
Unsuccessful attempt at trying to capture CRTL-C while cancelling a quickfix command

geekodour committed:
wip

justinmk commented:
how is quickfix related to #8200 ? I would think a change is needed in the area of `ui_puts` or `char_needs_redraw`.  I found those cases using `git grep` for `p_wd` (which is the 'writedelay' option):

    git grep p_wd

geekodour committed:
Added checks for force stop in ui_puts.

geekodour commented:
@justinmk oh! I thought It was the `ex_redraw` command that was causing the issue. Thanks for pointing out. 

The latest commit fixes the issue.

justinmk commented on src/nvim/ui.c:
```diff
@@ -441,8 +441,12 @@ void ui_puts(uint8_t *str)
+        uint64_t wd = (uint64_t)labs(p_wd);
+        os_delay(wd, false);
+      } else {
+        p_wd = 0;
```
what happens if you remove this? I think we would prefer not to modify 'writedelay' on ctrl-c, that would be a different kind of annoying ...

blueyed replied:
You then have to `Ctrl-C` multiple times, e.g. after `C-w s`,which I think is good, and similar to how Vim behaves.

bfredl replied:
what if a timer or something causes frequent redraws? At least a second `Ctrl-C` in a short while should reset `p_wd` if not the first. One can have keybindings that quickly sets `'wd'` to a sensible positive and negative value respectively.

blueyed replied:
Yes, I agree that a second `ctrl-c` in a short while should set it to 0.
If this is too complicated, then setting it on the first instance should do it probably (like in this PR), since not doing so makes it harder to abort due to the loop via the "Type  :quit<Enter>  to exit Nvim" message (see my last comment).

```diff
diff --git i/src/nvim/normal.c w/src/nvim/normal.c
index e4310de5d..385bfde57 100644
--- i/src/nvim/normal.c
+++ w/src/nvim/normal.c
@@ -1186,6 +1186,10 @@ static void normal_check_interrupt(NormalState *s)
       }
       got_int = false;
     }
+    // Reset "writedelay" on second interrupt.
+    if (s->previous_got_int) {
+      p_wd = 0;
+    }
     s->previous_got_int = true;
   } else {
     s->previous_got_int = false;
```

blueyed commented:
```diff
diff --git i/src/nvim/ui.c w/src/nvim/ui.c
index 4fb88718a..34c6bdd18 100644
--- i/src/nvim/ui.c
+++ w/src/nvim/ui.c
@@ -441,11 +441,9 @@ void ui_puts(uint8_t *str)
 
     if (p_wd) {  // 'writedelay': flush & delay each time.
       ui_flush();
-      if (!((got_int || (did_emsg && force_abort) || current_exception))){
+      if (!((got_int || (did_emsg && force_abort) || current_exception))) {
         uint64_t wd = (uint64_t)labs(p_wd);
         os_delay(wd, false);
-      } else {
-        p_wd = 0;
       }
     }
   }
```

blueyed commented on src/nvim/ui.c:
```diff
@@ -441,8 +441,12 @@ void ui_puts(uint8_t *str)
       ui_flush();
-      uint64_t wd = (uint64_t)labs(p_wd);
-      os_delay(wd, false);
+      if (!((got_int || (did_emsg && force_abort) || current_exception))){
```
Not sure about the other conditions besides `got_int` here: wouldn't that skip the delay in case of `echoerr` etc, which shouldn't be the case?

blueyed commented:
Could there be an indicator for the cell to be redrawn, possibly only with `&wd < 0`?
I've tried the following, but it does not work correctly:
```diff
diff --git i/src/nvim/ui.c w/src/nvim/ui.c
index 4fb88718a..4ac24d92e 100644
--- i/src/nvim/ui.c
+++ w/src/nvim/ui.c
@@ -424,6 +424,15 @@ void ui_puts(uint8_t *str)
     }
 
     size_t clen = (size_t)mb_ptr2len(p);
+
+    if (p_wd < 0) {
+      if (!((got_int || (did_emsg && force_abort) || current_exception))) {
+        ui_call_put((String){ .data = "X", .size = 1 });
+        ui_flush();
+        ui_cursor_goto(row, col);
+      }
+    }
+
     ui_call_put((String){ .data = (char *)p, .size = clen });
     col++;
     if (mb_ptr2cells(p) > 1) {
```

It might be even better to use highlights, e.g. toggle `reverse` before writing the new char.

blueyed commented:
Not setting `&wd = 0` makes it harder to abort, since `Ctrl-c` then triggers a redraw of the statusline / display of the "Type  :quit<Enter>  to exit Nvim" message (which needs to be canceled itself), and then continues with the previous redrawing (e.g. the syntax highlighting for the startify buffer).

blueyed commented:
The following patch on (top of this PR) seems to be even better:

It could additionally include a message for when `&wd` is reset on second interrupt.


```diff
 src/nvim/normal.c |  4 ++++
 src/nvim/screen.c | 12 ++++++++++++
 src/nvim/ui.c     | 10 ----------
 3 files changed, 16 insertions(+), 10 deletions(-)

diff --git i/src/nvim/normal.c w/src/nvim/normal.c
index e4310de5d..385bfde57 100644
--- i/src/nvim/normal.c
+++ w/src/nvim/normal.c
@@ -1186,6 +1186,10 @@ static void normal_check_interrupt(NormalState *s)
       }
       got_int = false;
     }
+    // Reset "writedelay" on second interrupt.
+    if (s->previous_got_int) {
+      p_wd = 0;
+    }
     s->previous_got_int = true;
   } else {
     s->previous_got_int = false;
diff --git i/src/nvim/screen.c w/src/nvim/screen.c
index 22de08041..d02f99708 100644
--- i/src/nvim/screen.c
+++ w/src/nvim/screen.c
@@ -5887,6 +5887,18 @@ static void screen_char(unsigned off, int row, int col)
   }
 
   ui_cursor_goto(row, col);
+
+  if (p_wd) {
+    if (!((got_int || (did_emsg && force_abort) || current_exception))) {
+      ui_set_highlight(ScreenAttrs[off] ^ HL_STANDOUT);
+      ui_putc('X');
+      ui_cursor_goto(row, col);
+      ui_flush();
+      uint64_t wd = (uint64_t)labs(p_wd);
+      os_delay(wd, false);
+    }
+  }
+
   ui_set_highlight(ScreenAttrs[off]);
 
   if (enc_utf8 && ScreenLinesUC[off] != 0) {
diff --git i/src/nvim/ui.c w/src/nvim/ui.c
index 4fb88718a..fbbf764de 100644
--- i/src/nvim/ui.c
+++ w/src/nvim/ui.c
@@ -438,16 +438,6 @@ void ui_puts(uint8_t *str)
       ui_linefeed();
     }
     p += clen;
-
-    if (p_wd) {  // 'writedelay': flush & delay each time.
-      ui_flush();
-      if (!((got_int || (did_emsg && force_abort) || current_exception))){
-        uint64_t wd = (uint64_t)labs(p_wd);
-        os_delay(wd, false);
-      } else {
-        p_wd = 0;
-      }
-    }
   }
 }
```

blueyed commented:
@geekodour 
Thanks for your feedback.  Please feel free to apply the patch to this PR.

What do you think about a message when `&wd` is reset?  (would include translations then, but maybe this could be kept english-only)

justinmk commented:
> What do you think about a message when &wd is reset? 

Let's just document it. 'writedelay' is only for developers anyways.

geekodour committed:
Updated docs for writedelay, moved p_wd handle part

Moved p_wd handler part to screen.c as suggested here: https://github.com/neovim/neovim/pull/8201#issuecomment-377221017

geekodour committed:
options.txt : Typo fix

geekodour commented:
I'll write a test for it.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -6926,4 +6926,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 
+	Once writedelay is set to some postitive or negative integer, it
+	can be reset to 0 by triggering two interrupts (Hitting <C-C> two
+	times)
```
Too verbose. Instead:

    <C-C><C-C> resets writedelay to 0.

blueyed replied:
Or `Using CTRL-C twice (during one main loop iteration) will reset it.`.

justinmk commented on src/nvim/normal.c:
```diff
@@ -1186,6 +1186,7 @@ static void normal_check_interrupt(NormalState *s)
       }
       got_int = false;
     }
+    if (s->previous_got_int) { p_wd = 0; }
```
body always goes on the next line.

```c
if (s->previous_got_int) {
  p_wd = 0;
}
```

justinmk commented on src/nvim/screen.c:
```diff
@@ -5887,6 +5887,18 @@ static void screen_char(unsigned off, int row, int col)
+    if (!got_int) {
+      ui_set_highlight(ScreenAttrs[off] ^ HL_STANDOUT);
+      ui_putc('X');
+      ui_cursor_goto(row, col);
```
ui_cursor_goto seems to hide the effects (if any) of  HL_STANDOUT/`ui_putc('X')`. The cursor placement helps  visually, so I don't think HL_STANDOUT/`ui_putc('X')` is needed.

Also, I tested this locally, and moved it back to `ui_puts()` and it seems to work fine there, so it should stay there.

The key change is the checks for `got_int` . Let's move forward with that.

blueyed replied:
> ui_cursor_goto seems to hide the effects (if any) of HL_STANDOUT/ui_putc('X')

Depends on the cursor shape (IIRC).
But I agree that it is not necessary really, and using a block cursor would be the better improvement (but is not necessary).

geekodour committed:
Moved interrupt check back to ui.c and changes to docs

geekodour committed:
wip

geekodour committed:
added test for canceling writedelay

blueyed commented:
@geekodour 
What's the state of this?  Would be great to get this finished - and I'd be happy to help.

bfredl commented:
Needs rebase. `ui_puts` was changed into `ui_line`, though the same logic should apply.

geekodour commented:
@blueyed @bfredl I am very rusty on the nvim codebase. I'll rebase this to resolve the conflicts today. I think it will be resolved faster if someone takes over the PR after that. :/
